### PR TITLE
MPS Implementation p1.

### PR DESCRIPTION
**Pull request by MichaelBroughton**
First step in adding the MPS design discussed in the simulator meetings. This PR adds the bare bones `mps_statespace.h` file. Once this is reviewed and in, the next step will be to put in `mps_simulator.h` (Which I'm planning to use Eigen3 for instead of writing out the intrinsics by hand) starting with single qubit functionality + tests and then building up to two qubit functionality + tests.

### CODE REVIEW

**Comment by MichaelBroughton** [comment]
@95-martin-orion Was that CI failure me or was it already breaking ?

File: lib/mps_statespace.h [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    auto end_sizes = 2 * 4 * bond_dim;
+    auto internal_sizes = 4 * bond_dim * bond_dim * num_qubits;
+    // use two extra "internal style" blocks past the end of the
+    //   working allocation for scratch space. Needed for gate
+    //   application.
+    auto size = sizeof(fp_type) * (end_sizes + internal_sizes);
+
+#ifdef _WIN32
+    Pointer ptr{(fp_type*)_aligned_malloc(size, 64), &detail::free};
+    bool is_null = ptr.get() != nullptr return MPS{std::move(ptr),
```
Comment (sergeisakov) [inline lib/mps_statespace.h:93]: Should the return statement be on a separate line?

File: lib/mps_statespace.h [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  // Get the pointer offset to the beginning of an MPS block.
+  unsigned GetBlockOffset(const MPS& state, unsigned i) const {
+    if (i == 0) {
+      return 0;
+    }
+    return 4 * state.bond_dim() * (1 + state.bond_dim() * (i - 1));
+  }
+
+  // Copies the state contents of one MPS to another.
+  //  ignores scratch data.
```
Comment (sergeisakov) [inline lib/mps_statespace.h:125]: Remove space and make "i" capital?

File: lib/mps_statespace.h [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  // Note: ForArgs are currently unused.
+  template <typename... ForArgs>
+  MPSStateSpace(ForArgs&&... args) : for_(args...) {}
+
+  // Requires num_qubits >= 2 and bond_dim >= 2.
+  static MPS CreateMPS(unsigned num_qubits, unsigned bond_dim) {
+    auto end_sizes = 2 * 4 * bond_dim;
+    auto internal_sizes = 4 * bond_dim * bond_dim * num_qubits;
+    // use two extra "internal style" blocks past the end of the
```
Comment (sergeisakov) [inline lib/mps_statespace.h:86]: Make 'u' capital?

File: lib/mps_statespace.h [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+#endif
+  }
+
+  unsigned Size(const MPS& state) const {
+    auto end_sizes = 2 * 4 * state.bond_dim();
+    auto internal_sizes = 4 * state.bond_dim() * state.bond_dim();
+    return end_sizes + internal_sizes * (state.num_qubits() - 2);
+  }
+
+  unsigned RawSize(const MPS& state) const {
```
Comment (sergeisakov) [inline lib/mps_statespace.h:112]: `unsigned` is sufficient to store all the sizes and offsets, right?

**Review by sergeisakov — CHANGES_REQUESTED** [review]
Looks good to me. I just have a few minor suggestions.

File: lib/mps_statespace.h (hunk 4 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h:112]: I think so. Rolling with unsigned types lets you address 2**32 - 1 places. Which is like 15+ Gigabytes worth of floats. I think the computational complexity of MPS being `n^3` for multiplications and SVDs vs `n^2` on memory you'd end up waiting an eternity to diagonalize such big matrices where memory wouldn't really be the main issue. Would you feel better if I bumped up to `long` ?

File: lib/mps_statespace.h (hunk 1 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Oops. Somehow that still worked with the missing semicolon. Fixed.

File: lib/mps_statespace.h (hunk 2 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Done.

File: lib/mps_statespace.h (hunk 3 above)
Comment (MichaelBroughton) [reply lib/mps_statespace.h]: Done.

File: lib/mps_statespace.h (hunk 4 above)
Comment (sergeisakov) [reply lib/mps_statespace.h:112]: If `unsigned` is sufficient then you don't need to use `long`.

**Review by sergeisakov — APPROVED** [review]
LGTM

**Pull request merged**